bootstrap/api.go

```diff
@@ -0,0 +1,58 @@
+package bootstrap
+
+import (
+	"fmt"
+	"os"
+
+	"github.com/buildkite/agent/v3/experiments"
+	"github.com/buildkite/agent/v3/jobapi"
+)
+
+// startJobAPI starts the job API server, iff the job API experiment is enabled, and the OS of the box supports it
+// otherwise it returns a noop cleanup function
+// It also sets the BUILDKITE_AGENT_JOB_API_SOCKET and BUILDKITE_AGENT_JOB_API_TOKEN environment variables
+func (b *Bootstrap) startJobAPI() (cleanup func(), err error) {
+	cleanup = func() {}
+
+	if !experiments.IsEnabled("job-api") {
+		return cleanup, nil
+	}
+
+	if !jobapi.Available() {
+		b.shell.Warningf("The Job API isn't available on this machine, as it's running an unsupported version of Windows")
+		b.shell.Warningf("The Job API is available on Unix agents, and agents running Windows versions after build 17063")
+		b.shell.Warningf("We'll continue to run your job, but you won't be able to use the Job API")
+		return cleanup, nil
+	}
+
+	socketPath, err := jobapi.NewSocketPath()
+	if err != nil {
+		return cleanup, fmt.Errorf("creating job API socket path: %v", err)
+	}
+
+	srv, token, err := jobapi.NewServer(socketPath, b.shell.Env)
```

Comment:
I think there's a problem here (similar to the missing `SOCKET` and`TOKEN` vars): `setUp` overwrites `b.shell.Env` with a new map, but because that's after the call to `startJobAPI`,`NewServer` gets (a reference to) the old map. Changes from `PATCH` and `DELETE` can be seen with `GET`, but won't be present as 100% Genuine environment variables in later hooks/plugins.